### PR TITLE
[CRT-399] Close the remaining kernel-dialog races in the teacher flow

### PR DESCRIPTION
**Pull request by pierluca**
## Summary

Teachers no longer get asked to select the Pyodide kernel when defining a task (Jira: CRT-399, jam e615a8d5 — "fixed for students, not for teachers").

## Background

The existing fix (`openTaskNotebook` awaiting kernelspecs) is mode-agnostic and already covers the common teacher path; the student flow was fixed by the same commit. The teacher-only repro came from two remaining openings that bypassed the guard — both exercised disproportionately by the teacher flow:

1. **Hidden grading panel** (`executeRunNotebookCommand`): a raw `documentManager.open` reached from the teacher's *save task* path (`getTask` → grading run). With specs not yet registered, JupyterLab's session init cannot resolve a default kernel and pops the dialog — over a hidden panel. Now awaits the same kernelspec wait; the extension's lint rule (which previously only banned `openOrReveal`) now also bans raw `documentManager.open`, verified to fire.
2. **Eager `sessionContext.initialize()`** (`preInstallPackages`): fired the moment a panel was added, racing `Context._populate`. When our call won, the session resolved its kernel *before* the notebook's `kernelspec` metadata reached the kernel preference → `getDefaultKernel` returns null → prompt, even with specs present. The background kernel start (for package pre-install) now waits for `context.ready`, after which it collapses into the context's own memoized `initialize` — the pre-install still runs in the background, just no longer ahead of the metadata.

## Refactor

`waitForKernelSpecs` moves out of the iframe API into `src/kernel/wait-for-kernel-specs.ts` (structural manager type) so both call sites share it — and it finally has unit coverage: specs already present, specs arriving late, empty emissions ignored, bounded give-up (still opens, warns), and a registration the subscription would have missed. Written red-first.

## Verification

Full notebook-runner suite 62/62; tsc + eslint + prettier clean. A teacher-flow e2e remains infeasible (saving a Jupyter task runs otter-grader in JupyterLite; the Pyodide kernel never finishes preparing in headless Chromium — same reason documented when the student e2e was written), so the races are covered by the unit tests + the lint gate.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Prevents the Pyodide kernel selection dialog for teachers by waiting for kernelspecs before any notebook opens and starting background kernels only after the notebook context is ready, with a bounded fallback and context-failure handling. Addresses CRT-399.

- **Bug Fixes**
  - Guard the hidden grading notebook open in `executeRunNotebookCommand` by awaiting `waitForKernelSpecs(app.serviceManager.kernelspecs)`; extend ESLint to also flag raw `documentManager.open`.
  - Start background kernels after `panel.context.ready` in `preInstallPackages`; handle failed context loads (log and skip) to avoid unhandled rejections.
  - Extract `waitForKernelSpecs` to `src/kernel/wait-for-kernel-specs.ts`: arm timeout/listener before readiness, resolve on any spec, and fall back on timeout or readiness reject/never-ready; add unit tests for late specs, empty emissions, reject/never-ready, and late spec after reject.

<sup>Written for commit a5743aeb2c57310ac5e4204fdd427cc160057979. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/crt25/collimator/pull/682?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * what the notebook's metadata says. The wait is bounded so that a broken
+ * kernel extension degrades to the pre-existing behavior (the notebook opens
+ * and JupyterLab may prompt for a kernel) instead of hanging the caller -
+ * and with it the embedded app - forever.
+ */
+export const waitForKernelSpecs = async (
+  kernelSpecs: KernelSpecWaitTarget,
+  timeoutMs: number = defaultTimeoutMs,
+): Promise<void> => {
+  await kernelSpecs.ready;
```
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts]: Fair enough!

File: apps/jupyter/extensions/notebook-runner/src/packages.ts [hunk 2]
```diff
@@ -115,17 +114,27 @@ const autoInstallPackages =
...
+      console.error(
+        `${logModule} Failed to initialize session for`,
+        notebookPath,
+        error,
+      );
+    }),
+  );
 
   return setupKernel(
     sessionContext,
```
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/packages.ts]: **Claude:**  The new void panel.context.ready.then(...) chain has no outer .catch: if context.ready rejects, that's a fresh unhandled rejection introduced by this PR — add a .catch with a log (one line). The second half ("stuck setupKernel, reject _packagesReady") describes the pre-existing failure behavior — before this PR, a failed initialize() also left setupKernel waiting — so I wouldn't restructure the flow in this PR.

File: apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  readonly ready: Promise<void>;
+  readonly specs: KernelSpec.ISpecModels | null;
+  readonly specsChanged: {
+    connect(callback: SpecsChangedCallback): unknown;
+    disconnect(callback: SpecsChangedCallback): unknown;
+  };
+}
+
+/**
+ * Resolves once at least one kernelspec (the Pyodide kernel) is registered,
```
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts]: **Claude:** pedantic but cheap. The parenthetical is a statement about this app (its only kernel is Pyodide), not a claim about the check; the text also predates the PR (it moved files). Reword to "at least one kernelspec — in this app that's the Pyodide kernel —" and both bots go quiet. Optional.

File: apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts]: Fixed in 0e669537. waitForKernelSpecs now arms the timeout and the specsChanged listener *before* touching `kernelSpecs.ready`, and treats a resolved, rejected, or never-settling readiness as the same bounded fallback. Three new tests cover it: readiness rejects, readiness never settles, and a late spec arriving after a rejected readiness.

File: apps/jupyter/extensions/notebook-runner/src/packages.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/packages.ts]: Fixed in 0e669537, per the scope you set: added a one-line `.catch` that logs the failed context load (the await is now `panel.context.ready.catch(...)`), so it is no longer an unhandled rejection. I left the flow otherwise unchanged — the stuck-setupKernel-on-failed-load behavior predates this PR, so not restructured here.

File: apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts]: Fixed in 0e669537. Reworded to "at least one kernelspec — in this app that is the Pyodide kernel —", so it no longer implies the check is Pyodide-specific.

[the remark this answers is not in the document]
Comment (pierluca) [reply apps/jupyter/extensions/notebook-runner/src/kernel/wait-for-kernel-specs.ts]: Fixed in 0e669537 — reworded the docstring to "at least one kernelspec — in this app that is the Pyodide kernel —" so it no longer implies a stricter, Pyodide-specific check than `hasSpec` performs.

**Review by bubudetp — APPROVED** [review]
